### PR TITLE
Adds seven new crew objectives for various jobs

### PR DESCRIPTION
**Pull request by 112358sam**
[feature]

## About the PR 
Adds seven new crew objectives. Two objectives for the engineer and CE, two for the bartender, two for the chef, and one for the geneticist. Also gives the engineer the CE objective for all furnaces on at the end of the round.

One of the bartender objectives and both chef objectives randomly generate the required food from the list of possible food or drink items. The list is filtered to remove impossible or admin-only foods and drinks.

Engineer and CE:
Make sure all SMES units on the station are at least 20% charged at the end of the round.
Ensure all APC units on the station are operating at the end of the round.

Bartender
Don't lose your shotgun!
Mix three randomly selected drinks using your cocktail shaker.

Chef
Create a custom, three-tier cake with three randomly selected food infused layers in any order.
Create a custom pizza with three randomly selected food toppings.

Geneticist
Have at least 5 options available in the gene booth at the end of the round.

## Why's this needed? 
With this PR, all common jobs will have at least one crew objective with most having at least two. It gives a low-stakes goal for players to pursue and the goals should encourage players to learn more about their job and complete some of the basic tasks. I feel like the randomly generated goals for the chef and bartender would be particularly fun since they may require the player to branch out and make food or drinks they may have not had any reason to in the past.

## Changelog

```changelog
(u)112358sam
(*)Added seven new crew objectives.
(+)Added the furnace objective that the CE has to the possible engineer objectives.
```

### CODE REVIEW

**Comment by 112358sam** [comment]
> 
> 
> As for the geneticist objective, I'm wondering if it's more interesting to keep a tally of gene booth sales instead?

Personally I feel that takes it too much out of the geneticists hands. On low pop that objective can become much harder and it would still be game-able by a geneticist who just buys their own genes. I think your idea does tie it much closer to actually helping the rest of the crew.

File: code/datums/crew_objective.dm [hunk 1]
```diff
@@ -197,6 +233,12 @@ ABSTRACT_TYPE(/datum/objective/crew/quartermaster)
 			if(wagesystem.shipping_budget > 50000) return 1
 			else return 0
 
+	specialorder
+		explanation_text = "Fulfill an off-station order requisition."
+		var/completed = 0
+		check_completion()
+			return length(shippingmarket.complete_orders)
+
```
[the remark this answers is not in the document]
Comment (112358sam) [reply code/datums/crew_objective.dm:233]: PR merged and conflicts resolved

File: code/datums/crew_objective.dm [hunk 2]
```diff
@@ -147,9 +147,9 @@ ABSTRACT_TYPE(/datum/objective/crew/chiefengineer)
 		medal_name = "Slow Burn"
 		check_completion()
 			for(var/obj/machinery/power/furnace/F in machine_registry[MACHINES_POWER])
-				if(F.z == 1 && F.active == 1)
-					return 1
-			return 0
+				if(F.z == 1 && F.active == 0 && !istype(F.loc.loc, /area/listeningpost))
+					return 0
```
[the remark this answers is not in the document]
Comment (112358sam) [reply code/datums/crew_objective.dm]: Checking for in /area/station subtype now. Hopefully that is more robust.

File: code/datums/crew_objective.dm [hunk 3]
```diff
@@ -159,6 +159,42 @@ ABSTRACT_TYPE(/datum/objective/crew/chiefengineer)
...
+					if(PN.nodes[1].z != 1)
+						continue
+					if(!stationnet || length(PN.nodes) > length(stationnet.nodes))
+						stationnet = PN
+
+			for(var/obj/machinery/power/terminal/term in stationnet.nodes)
+				var/obj/machinery/power/apc/A = term.master
+				if(istype(A) && (!A.area || A.area.requires_power))
+					if(!A.operating)
+						return 0
```
[the remark this answers is not in the document]
Comment (112358sam) [reply code/datums/crew_objective.dm]: Done, and this approach adapted to other objectives.

File: code/datums/crew_objective.dm [hunk 4]
```diff
@@ -159,6 +159,42 @@ ABSTRACT_TYPE(/datum/objective/crew/chiefengineer)
...
+		explanation_text = "Ensure all APC units on the station are operating at the end of the round."
+		check_completion()
+			//Find the station powernet - largest z1 powernet
+			var/datum/powernet/stationnet = null
+			for (var/datum/powernet/PN as anything in powernets)
+				if(length(PN.nodes))
+					if(PN.nodes[1].z != 1)
+						continue
+					if(!stationnet || length(PN.nodes) > length(stationnet.nodes))
+						stationnet = PN
```
[the remark this answers is not in the document]
Comment (112358sam) [reply code/datums/crew_objective.dm]: Fixed by switch to area check.

File: code/datums/crew_objective.dm [hunk 5]
```diff
@@ -159,6 +159,42 @@ ABSTRACT_TYPE(/datum/objective/crew/chiefengineer)
...
 
+	reserves
+		explanation_text = "Make sure all SMES units on the station are at least half charged at the end of the round."
+		check_completion()
+			//Find the station powernet - largest z1 powernet
+			var/datum/powernet/stationnet = null
+			for (var/datum/powernet/PN as anything in powernets)
+				if(length(PN.nodes))
+					if(PN.nodes[1].z != 1)
+						continue
```
[the remark this answers is not in the document]
Comment (112358sam) [reply code/datums/crew_objective.dm]: Fixed by switch to area check

File: code/datums/crew_objective.dm [hunk 6]
```diff
@@ -159,6 +159,23 @@ ABSTRACT_TYPE(/datum/objective/crew/chiefengineer)
...
+		check_completion()
+			for (var/obj/machinery/power/smes/S in machine_registry[MACHINES_POWER])
+				if(istype(get_area(S),/area/station) && S.charge < S.capacity/5)
+					return 0
+			return 1
+
+	apc
+		explanation_text = "Ensure all APC units on the station are operating at the end of the round."
+		check_completion()
+			for(var/obj/machinery/power/apc/A in by_type[/obj/machinery/power/apc])
```
Comment (Azrun) [inline code/datums/crew_objective.dm:173]: consider using `for_by_tcl`
```for_by_tcl(A, /obj/machinery/power/apc)```

File: code/datums/crew_objective.dm [hunk 7]
```diff
@@ -208,6 +225,12 @@ ABSTRACT_TYPE(/datum/objective/crew/quartermaster)
 			if(wagesystem.shipping_budget > 50000) return 1
 			else return 0
 
+	specialorder
+		explanation_text = "Fulfill an off-station order requisition."
+		var/completed = 0
```
Comment (Azrun) [inline code/datums/crew_objective.dm:230]:
```suggestion
		var/completed = FALSE
```

File: code/datums/crew_objective.dm [hunk 8]
```diff
@@ -307,11 +330,143 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
...
+#define DRINK_OBJ_3 4
+ABSTRACT_TYPE(/datum/objective/crew/bartender)
+/datum/objective/crew/bartender
+	shotgun
+		explanation_text = "Don't lose your shotgun!"
+		check_completion()
+			if(owner.current && owner.current.check_contents_for(/obj/item/gun/kinetic/riotgun))
+				return 1
+			else
+				return 0
```
Comment (Azrun) [inline code/datums/crew_objective.dm:345]:
```suggestion
				return TRUE
			else
				return FALSE
```

File: code/datums/crew_objective.dm [hunk 9]
```diff
@@ -307,11 +330,143 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
...
+	drinks
+		var/completed = 0
```
Comment (Azrun) [inline code/datums/crew_objective.dm:347]:
```suggestion
		var/completed = FALSE
```

File: code/datums/crew_objective.dm [hunk 10]
```diff
@@ -307,11 +330,143 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
 					return 0
 			return 1
 
-//	bartender
+#define DRINK_OBJ_DONE 7
+#define DRINK_OBJ_1 1
+#define DRINK_OBJ_2 2
+#define DRINK_OBJ_3 4
```
Comment (Azrun) [inline code/datums/crew_objective.dm:336]: Define the number of drinks as well, you reference `3` in a bunch of places for `/datum/objective/crew/bartender/drinks`

File: code/datums/crew_objective.dm [hunk 11]
```diff
@@ -307,11 +330,143 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
...
+		var/choices[3]
+		var/completed = 0
+		New()
+			..()
+			//Result is cached so it should just access the list after the first time
+			var/list/ingredients = concrete_typesof(/obj/item/reagent_containers/food/snacks)
+			ingredients -= src.blacklist
+			ingredients -= concrete_typesof(/obj/item/reagent_containers/food/snacks/ingredient/egg/critter)
+			var/list/names[3]
+			for(var/i = 1, i <= 3, i++)
```
Comment (Azrun) [inline code/datums/crew_objective.dm:410]: Same issue with 3, use a define so `choices`, `names` and iterating over them are coupled.

File: code/datums/crew_objective.dm [hunk 12]
```diff
@@ -307,11 +330,143 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
...
+	pizza
+		var/choices[3]
+		var/completed = 0
+		New()
+			..()
+			var/list/ingredients = concrete_typesof(/obj/item/reagent_containers/food/snacks)
+			ingredients -= src.blacklist
+			ingredients -= concrete_typesof(/obj/item/reagent_containers/food/snacks/ingredient/egg/critter)
+			var/list/names[3]
+			for(var/i = 1, i <= 3, i++)
```
Comment (Azrun) [inline code/datums/crew_objective.dm:431]: Same issue with 3, use a define so choices, names and iterating over them are coupled.

File: code/datums/crew_objective.dm [hunk 13]
```diff
@@ -307,11 +330,143 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
...
+		check_completion()
+			for (var/obj/machinery/power/smes/S in machine_registry[MACHINES_POWER])
+				if(istype(get_area(S),/area/station) && S.charge < S.capacity/5)
+					return FALSE
+			return TRUE
+
+	apc
+		explanation_text = "Ensure all APC units on the station are operating at the end of the round."
+		check_completion()
+			for(var/obj/machinery/power/apc/A in by_type[/obj/machinery/power/apc])
```
Comment (Azrun) [inline code/datums/crew_objective.dm:465]: `for_by_tcl`

File: code/modules/food_and_drink/cakes.dm [hunk 14]
```diff
@@ -291,6 +292,14 @@
 		if(c.litfam)
 			src.ignite()
 		src.update_cake_context()
+
+		//Complete pizza crew objectives if possible
```
Comment (Azrun) [inline code/modules/food_and_drink/cakes.dm:296]:
```suggestion
		//Complete cake crew objectives if possible
```

File: code/obj/submachine/cooking.dm [hunk 15]
```diff
@@ -643,6 +643,16 @@ table#cooktime a#start {
...
 					output = R.specialOutput(src)
+
+					//Complete pizza crew objectives if possible
+					if(istype(output,/obj/item/reagent_containers/food/snacks/pizza/))
+						var/obj/item/reagent_containers/food/snacks/pizza/P = output
+						if (usr.mind && usr.mind.objectives)
+							for (var/datum/objective/crew/chef/pizza/O in usr.mind.objectives)
+								var/list/matching_toppings = P.topping_types & O.choices
+								if(matching_toppings.len >= 3)
+									O.completed = 1
```
Comment (Azrun) [inline code/obj/submachine/cooking.dm:654]:
```suggestion
									O.completed = TRUE
```

**Review by Azrun — CHANGES_REQUESTED** [review]

File: code/datums/crew_objective.dm (hunk 9 above)
Comment (112358sam) [reply code/datums/crew_objective.dm]: Done, thanks!

File: code/datums/crew_objective.dm (hunk 10 above)
Comment (112358sam) [reply code/datums/crew_objective.dm]: Done!

File: code/datums/crew_objective.dm (hunk 11 above)
Comment (112358sam) [reply code/datums/crew_objective.dm]: Done

File: code/obj/submachine/cooking.dm [hunk 16]
```diff
@@ -643,6 +643,16 @@ table#cooktime a#start {
 						if (!OVEN_checkitem(R.item4, R.amt4)) continue
 
 					output = R.specialOutput(src)
+
+					//Complete pizza crew objectives if possible
+					if(istype(output,/obj/item/reagent_containers/food/snacks/pizza/))
+						var/obj/item/reagent_containers/food/snacks/pizza/P = output
+						if (usr.mind && usr.mind.objectives)
```
Comment (ZeWaka) [inline code/obj/submachine/cooking.dm:650]:
```suggestion
						if (usr.mind?.objectives)
```

File: code/obj/submachine/cooking.dm [hunk 17]
```diff
@@ -643,6 +643,16 @@ table#cooktime a#start {
...
+							for (var/datum/objective/crew/chef/pizza/O in usr.mind.objectives)
+								var/list/matching_toppings = P.topping_types & O.choices
+								if(matching_toppings.len >= PIZZA_OBJ_COUNT)
+									O.completed = TRUE
```
Comment (ZeWaka) [inline code/obj/submachine/cooking.dm:654]: bad variable name. bad.
bad `len`. bad.
```suggestion
							for (var/datum/objective/crew/chef/pizza/objective in usr.mind.objectives)
								var/list/matching_toppings = P.topping_types & objective.choices
								if(length(matching_toppings) >= PIZZA_OBJ_COUNT)
									objective.completed = TRUE
```

File: code/modules/food_and_drink/snacks.dm [hunk 18]
```diff
@@ -72,6 +72,7 @@
 	var/sliced = FALSE
 	var/topping = FALSE
 	var/num = 0
+	var/list/topping_types = list()
```
Comment (ZeWaka) [inline code/modules/food_and_drink/snacks.dm:75]: unused?

File: code/modules/food_and_drink/cakes.dm [hunk 19]
```diff
@@ -291,6 +292,14 @@
...
 			src.ignite()
 		src.update_cake_context()
+
+		//Complete cake crew objectives if possible
+		src.cake_types += c.cake_types
+		if (user.mind && user.mind.objectives)
+			for (var/datum/objective/crew/chef/cake/O in user.mind.objectives)
+				var/list/matching_types = src.cake_types & O.choices
+				if(matching_types.len >= CAKE_OBJ_COUNT)
+					O.completed = TRUE
```
Comment (ZeWaka) [inline code/modules/food_and_drink/cakes.dm:302]: same bad.

File: code/datums/crew_objective.dm [hunk 20]
```diff
@@ -562,6 +735,14 @@ ABSTRACT_TYPE(/datum/objective/crew/geneticist)
 					return 1
 			return 0
 
+	booth
+		explanation_text = "Have at least 5 options available in the gene booth at the end of the round."
+		check_completion()
+			var/list/geneoptions = list()
+			for_by_tcl(GB, /obj/machinery/genetics_booth)
+				geneoptions |= GB.offered_genes
+			return geneoptions.len >= 5
```
Comment (ZeWaka) [inline code/datums/crew_objective.dm:744]:
```suggestion
			return length(geneoptions) >= 5
```

File: code/datums/crew_objective.dm [hunk 21]
```diff
@@ -307,11 +330,161 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
...
+			return completed
+
+ABSTRACT_TYPE(/datum/objective/crew/engineer)
+/datum/objective/crew/engineer
+	furnaces
+		explanation_text = "Make sure all furnaces on the station are active at the end of the round."
+		medal_name = "Slow Burn"
+		check_completion()
+			for(var/obj/machinery/power/furnace/F in machine_registry[MACHINES_POWER])
+				if(F.z == 1 && !F.active && istype(get_area(F), /area/station))
```
Comment (ZeWaka) [inline code/datums/crew_objective.dm:468]: 1 bad. station z define good.

File: code/datums/crew_objective.dm [hunk 22]
```diff
@@ -147,7 +147,7 @@ ABSTRACT_TYPE(/datum/objective/crew/chiefengineer)
 		medal_name = "Slow Burn"
 		check_completion()
 			for(var/obj/machinery/power/furnace/F in machine_registry[MACHINES_POWER])
-				if(F.z == 1 && !F.active && istype(F.loc.loc, /area/station))
+				if(F.z == 1 && !F.active && istype(get_area(F), /area/station))
```
Comment (ZeWaka) [inline code/datums/crew_objective.dm:150]: ditto

File: code/datums/crew_objective.dm [hunk 23]
```diff
@@ -208,6 +225,12 @@ ABSTRACT_TYPE(/datum/objective/crew/quartermaster)
 			if(wagesystem.shipping_budget > 50000) return 1
 			else return 0
 
+	specialorder
+		explanation_text = "Fulfill an off-station order requisition."
+		var/completed = FALSE
```
Comment (ZeWaka) [inline code/datums/crew_objective.dm:230]: wat

File: code/datums/crew_objective.dm [hunk 24]
```diff
@@ -307,11 +330,161 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
 					return 0
 			return 1
 
-//	bartender
+#define DRINK_OBJ_COUNT 3
+ABSTRACT_TYPE(/datum/objective/crew/bartender)
+/datum/objective/crew/bartender
+	shotgun
```
Comment (ZeWaka) [inline code/datums/crew_objective.dm:336]: ideally these would not be relatively pathed like this.

File: code/datums/crew_objective.dm [hunk 25]
```diff
@@ -307,11 +330,161 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
...
+		/obj/item/reagent_containers/food/snacks/ingredient/meat/mysterymeat/nugget/flock,
+		/obj/item/reagent_containers/food/snacks/ingredient/pepperoni,
+		/obj/item/reagent_containers/food/snacks/meatball,
+		/obj/item/reagent_containers/food/snacks/mushroom,
+		/obj/item/reagent_containers/food/snacks/pickle/trash,
+		/obj/item/reagent_containers/food/snacks/pizza/xmas,
+		/obj/item/reagent_containers/food/snacks/plant,
+		/obj/item/reagent_containers/food/snacks/plant/glowfruit/spawnable,
+	)
+	cake
```
Comment (ZeWaka) [inline code/datums/crew_objective.dm:406]: ideally these would not be relatively pathed like this.

File: code/datums/crew_objective.dm [hunk 26]
```diff
@@ -307,11 +330,161 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
...
 			return 1
 
-//	bartender
+#define DRINK_OBJ_COUNT 3
+ABSTRACT_TYPE(/datum/objective/crew/bartender)
+/datum/objective/crew/bartender
+	shotgun
+		explanation_text = "Don't lose your shotgun!"
+		check_completion()
+			if(owner.current && owner.current.check_contents_for(/obj/item/gun/kinetic/riotgun))
```
Comment (ZeWaka) [inline code/datums/crew_objective.dm:339]:
```suggestion
			if(owner.current?.check_contents_for(/obj/item/gun/kinetic/riotgun))
```

File: code/datums/crew_objective.dm [hunk 27]
```diff
@@ -307,11 +330,161 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
...
+				/datum/reagent/fooddrink/alcoholic/rum,
+				/datum/reagent/fooddrink/alcoholic/tequila,
+				/datum/reagent/fooddrink/alcoholic/vermouth,
+				/datum/reagent/fooddrink/alcoholic/vodka,
+				/datum/reagent/fooddrink/alcoholic/wine,
+				/datum/reagent/fooddrink/alcoholic/wine/white
+			)
+			cocktails -= blacklist
+			var/list/names[DRINK_OBJ_COUNT]
+			for(var/i = 1, i <= DRINK_OBJ_COUNT, i++)
```
Comment (ZeWaka) [inline code/datums/crew_objective.dm:373]: for-in-to please

File: code/datums/crew_objective.dm [hunk 28]
```diff
@@ -307,11 +330,161 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
...
+				return TRUE
+			else
+				return FALSE
+	drinks
+		var/completed = FALSE
+		var/ids[DRINK_OBJ_COUNT]
+		New()
+			..()
+			var/list/cocktails = concrete_typesof(/datum/reagent/fooddrink/alcoholic)
+			var/list/blacklist = list(
```
Comment (ZeWaka) [inline code/datums/crew_objective.dm:349]: this can be static and on the type level

File: code/datums/crew_objective.dm [hunk 29]
```diff
@@ -307,11 +330,161 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
...
+		var/choices[CAKE_OBJ_COUNT]
+		var/completed = FALSE
+		New()
+			..()
+			//Result is cached so it should just access the list after the first time
+			var/list/ingredients = concrete_typesof(/obj/item/reagent_containers/food/snacks)
+			ingredients -= src.blacklist
+			ingredients -= concrete_typesof(/obj/item/reagent_containers/food/snacks/ingredient/egg/critter)
+			var/list/names[CAKE_OBJ_COUNT]
+			for(var/i = 1, i <= CAKE_OBJ_COUNT, i++)
```
Comment (ZeWaka) [inline code/datums/crew_objective.dm:416]: ditto

File: code/datums/crew_objective.dm [hunk 30]
```diff
@@ -307,11 +330,161 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
...
+			explanation_text += "using your cocktail shaker."
+
+		check_completion()
+			return completed == 2**DRINK_OBJ_COUNT-1 //Uses bit flags
+
+#define CAKE_OBJ_COUNT 3
+#define PIZZA_OBJ_COUNT 3
+ABSTRACT_TYPE(/datum/objective/crew/chef)
+/datum/objective/crew/chef
+	var/list/blacklist = list(
```
Comment (ZeWaka) [inline code/datums/crew_objective.dm:393]: static

File: code/datums/crew_objective.dm [hunk 31]
```diff
@@ -307,11 +330,161 @@ ABSTRACT_TYPE(/datum/objective/crew/janitor)
...
+	pizza
+		var/choices[PIZZA_OBJ_COUNT]
+		var/completed = FALSE
+		New()
+			..()
+			var/list/ingredients = concrete_typesof(/obj/item/reagent_containers/food/snacks)
+			ingredients -= src.blacklist
+			ingredients -= concrete_typesof(/obj/item/reagent_containers/food/snacks/ingredient/egg/critter)
+			var/list/names[PIZZA_OBJ_COUNT]
+			for(var/i = 1, i <= PIZZA_OBJ_COUNT, i++)
```
Comment (ZeWaka) [inline code/datums/crew_objective.dm:443]: ditto

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/modules/food_and_drink/snacks.dm (hunk 18 above)
Comment (112358sam) [reply code/modules/food_and_drink/snacks.dm:75]: It is used in the check to see if the pizza has the objective-requested toppings.

File: code/modules/food_and_drink/cakes.dm (hunk 19 above)
Comment (112358sam) [reply code/modules/food_and_drink/cakes.dm]: Fixed

File: code/datums/crew_objective.dm (hunk 21 above)
Comment (112358sam) [reply code/datums/crew_objective.dm]: Fixed

File: code/datums/crew_objective.dm (hunk 22 above)
Comment (112358sam) [reply code/datums/crew_objective.dm]: Fixed

File: code/datums/crew_objective.dm (hunk 23 above)
Comment (112358sam) [reply code/datums/crew_objective.dm]: Fixed

File: code/datums/crew_objective.dm (hunk 27 above)
Comment (112358sam) [reply code/datums/crew_objective.dm]: Fixed

File: code/datums/crew_objective.dm (hunk 28 above)
Comment (112358sam) [reply code/datums/crew_objective.dm]: Fixed

File: code/datums/crew_objective.dm (hunk 29 above)
Comment (112358sam) [reply code/datums/crew_objective.dm]: Fixed

File: code/datums/crew_objective.dm (hunk 30 above)
Comment (112358sam) [reply code/datums/crew_objective.dm]: Fixed

File: code/datums/crew_objective.dm (hunk 31 above)
Comment (112358sam) [reply code/datums/crew_objective.dm:436]: Fixed

File: code/datums/crew_objective.dm (hunk 24 above)
Comment (112358sam) [reply code/datums/crew_objective.dm]: fixed

File: code/datums/crew_objective.dm (hunk 25 above)
Comment (112358sam) [reply code/datums/crew_objective.dm]: fixed

**Comment by 112358sam** [comment]
I've updated the file to always use absolute paths for objective types. This did mean nearly every line was moved a tab leading to the increased file size.

**Pull request merged**